include/dice/template-library/polymorphic_allocator.hpp

```diff
@@ -314,33 +318,91 @@ namespace dice::template_library {
+	struct offset_ptr_stl_allocator {
 		using value_type = T;
 		using pointer = boost::interprocess::offset_ptr<T>;
+		using const_pointer = boost::interprocess::offset_ptr<T const>;
+		using void_pointer = boost::interprocess::offset_ptr<void>;
+		using const_void_pointer = boost::interprocess::offset_ptr<void const>;
 		using size_type = size_t;
 		using difference_type = std::ptrdiff_t;
+		using upstream_allocator_type = Allocator<T>;
 
-		using propagate_on_container_copy_assignment = typename std::allocator_traits<std::allocator<T>>::propagate_on_container_copy_assignment;
-		using propagate_on_container_move_assignment = typename std::allocator_traits<std::allocator<T>>::propagate_on_container_move_assignment;
-		using propagate_on_container_swap = typename std::allocator_traits<std::allocator<T>>::propagate_on_container_swap;
-		using is_always_equal = typename std::allocator_traits<std::allocator<T>>::is_always_equal;
+		using propagate_on_container_copy_assignment = typename std::allocator_traits<upstream_allocator_type>::propagate_on_container_copy_assignment;
+		using propagate_on_container_move_assignment = typename std::allocator_traits<upstream_allocator_type>::propagate_on_container_move_assignment;
+		using propagate_on_container_swap = typename std::allocator_traits<upstream_allocator_type>::propagate_on_container_swap;
+		using is_always_equal = typename std::allocator_traits<upstream_allocator_type>::is_always_equal;
 
-		constexpr offset_ptr_stl_allocator() noexcept = default;
+	private:
+		template<typename, template<typename> typename>
+		friend struct offset_ptr_stl_allocator;
+
+		[[no_unique_address]] upstream_allocator_type inner_;
+
+	public:
+		constexpr offset_ptr_stl_allocator() noexcept(std::is_nothrow_default_constructible_v<upstream_allocator_type>) = default;
+
+		constexpr offset_ptr_stl_allocator(offset_ptr_stl_allocator const &other) noexcept(std::is_nothrow_copy_constructible_v<upstream_allocator_type>)
+			: inner_{other.inner_} {
+		}
+
+		constexpr offset_ptr_stl_allocator(offset_ptr_stl_allocator &&other) noexcept(std::is_nothrow_move_constructible_v<upstream_allocator_type>)
+			: inner_{other.inner_} {
+		}
 
 		template<typename U>
-		constexpr offset_ptr_stl_allocator(offset_ptr_stl_allocator<U> const &) noexcept {}
+		constexpr offset_ptr_stl_allocator(offset_ptr_stl_allocator<U, Allocator> const &other) noexcept
+			: inner_{other.inner_} {
+		}
+
+		constexpr offset_ptr_stl_allocator &operator=(offset_ptr_stl_allocator const &other) noexcept(std::is_nothrow_copy_assignable_v<upstream_allocator_type>) = default;
+		constexpr offset_ptr_stl_allocator &operator=(offset_ptr_stl_allocator &&other) noexcept(std::is_nothrow_move_assignable_v<upstream_allocator_type>) = default;
+
+		explicit constexpr offset_ptr_stl_allocator(upstream_allocator_type const &upstream) noexcept(std::is_nothrow_copy_constructible_v<upstream_allocator_type>)
+			: inner_{upstream} {
```

Comment:
only activate if upstream_allocator_type is copy constructible.

Reply:
Are there allocators that are not copy constructible? Would that work? I think all containers assume that they are copy constructible.

The only thing I'm aware of is allocators that are not assignable or swappable

Reply:
- There are only constants to customize assignment and swap behaviour
- There is also `select_on_container_copy_construction` which is required to create a new instance.
- all container ctors take allocator by `const &`

Reply:
about copy construction: you are right "(Note: Every Allocator also satisfies [CopyConstructible](https://en.cppreference.com/w/cpp/named_req/CopyConstructible).)" from https://en.cppreference.com/w/cpp/named_req/Allocator

Reply:
reviewing the link above I come to the conclusion that every allocator must have move/copy assignment and copy ctor. I couldn't find anything about move ctor but I think it is safe to assume that it will always be present if there is already a move assignment operator.

Reply:
You are wrong about assignment, there is nothing about assignment being required on that page you sent me. Only move+copy ctors.

copy ctor
![grafik](https://github.com/user-attachments/assets/b29f14ae-80d5-474a-a471-064a6abef8a6)

move ctor
![grafik](https://github.com/user-attachments/assets/b3e2005d-30c4-474a-bcc9-2774fabdb57b)

This can also be confirmed by looking at [`std::pmr::polymorphic_allocator`](https://en.cppreference.com/w/cpp/memory/polymorphic_allocator) as it does not have any `operator=`.

Reply:
But I've now defaulted the copy and move ctor, as it does not make sense to implement them manually